openmls/src/group/tests/external_proposal.rs

```diff
@@ -0,0 +1,323 @@
+                bob_group.store_pending_proposal(*proposal)
+            }
+            _ => unreachable!(),
+        }
+
+        // and Alice will commit it
+        let (commit, welcome) = alice_group.commit_to_pending_proposals(backend).unwrap();
+        alice_group.merge_pending_commit().unwrap();
+        assert_eq!(alice_group.members().len(), 3);
+
+        // Bob will also process the commit
+        let parsed = bob_group.parse_message(commit.into(), backend).unwrap();
+        let msg = bob_group
+            .process_unverified_message(parsed, None, backend)
+            .unwrap();
+        match msg {
+            ProcessedMessage::StagedCommitMessage(commit) => {
+                bob_group.merge_staged_commit(*commit).unwrap()
+            }
+            _ => unreachable!(),
+        }
+        assert_eq!(bob_group.members().len(), 3);
+
+        // Finally, Charlie can join with the Welcome
+        let cfg = MlsGroupConfig::builder().wire_format_policy(policy).build();
+        let charlie_group = MlsGroup::new_from_welcome(
+            backend,
+            &cfg,
+            welcome.unwrap(),
+            Some(alice_group.export_ratchet_tree()),
+        )
+        .unwrap();
+        assert_eq!(charlie_group.members().len(), 3);
+    }
+}
+
+#[apply(ciphersuites_and_backends)]
+fn external_add_proposal_should_be_signed_by_key_package_it_references(
+    ciphersuite: Ciphersuite,
+    backend: &impl OpenMlsCryptoProvider,
+) {
+    let ProposalValidationTestSetup {
+        mut alice_group, ..
+    } = validation_test_setup(PURE_PLAINTEXT_WIRE_FORMAT_POLICY, ciphersuite, backend);
+
+    let attacker_cb = get_credential_bundle(
+        "Attacker".into(),
+        CredentialType::Basic,
+        ciphersuite.signature_algorithm(),
+        backend,
+    )
+    .unwrap();
+
+    // A new client, Charlie, will now ask joining with an external Add proposal
+    let charlie_cb = get_credential_bundle(
+        "Charlie".into(),
+        CredentialType::Basic,
+        ciphersuite.signature_algorithm(),
+        backend,
+    )
+    .unwrap();
+
+    let charlie_kp =
+        generate_key_package_bundle(&[ciphersuite], charlie_cb.credential(), vec![], backend)
+            .unwrap();
+
+    let invalid_proposal = JoinProposal::new(
+        charlie_kp,
+        alice_group.group_id().clone(),
+        alice_group.epoch(),
+        &attacker_cb,
+        backend,
+    )
+    .unwrap();
+
+    // fails because the message was not signed by the same credential as the one in the Add proposal
+    let invalid_msg = alice_group
+        .parse_message(invalid_proposal.into(), backend)
+        .unwrap();
+    assert!(matches!(
+        alice_group
+            .process_unverified_message(invalid_msg, None, backend)
+            .unwrap_err(),
+        UnverifiedMessageError::InvalidSignature
+    ));
+}
+
+#[apply(ciphersuites_and_backends)]
+fn new_member_proposal_sender_should_be_reserved_for_join_proposals(
```

Comment:
I think we will need semantic validation for sender types once we have all external proposals, but it's good to have these tests for now.

Reply:
Agree. I added a TODO to remind us later